{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "from random import randrange\n",
    "from time import time "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([[], [1, 2, 3, 4, 5, 6], [7, 8, 9], []], [0, 7, 11])"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def split_at_values(lst, values):\n",
    "                indices = [i for i, x in enumerate(lst) if x in values]\n",
    "                size = len(lst) \n",
    "                res = [lst[i: j] for i, j in zip([0] + [x+1 for x in indices], indices + ([size] if indices[-1] != size else []))]\n",
    "                return res, indices\n",
    "            \n",
    "split_at_values([\"A\",1,2,3,4,5,6,\"A\",7,8,9, \"A\"], [\"A\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#=========================================================================== GENETIC ALGORITHM =======================================\n",
    "# Class to represent problems to be solved by means of a general\n",
    "# genetic algorithm. It includes the following attributes:\n",
    "# - genes: list of possible genes in a chromosome\n",
    "# - individuals_length: length of each chromosome\n",
    "# - decode: method that receives the genotype (chromosome) as input and returns\n",
    "#    the phenotype (solution to the original problem represented by the chromosome) \n",
    "# - fitness: method that returns the evaluation of a chromosome (acts over the\n",
    "#    genotype)\n",
    "# - mutation: function that implements a mutation over a chromosome\n",
    "# - crossover: function that implements the crossover operator over two chromosomes\n",
    "#=====================================================================================================================================\n",
    "\n",
    "class Problem_Genetic(object):\n",
    "    \n",
    "    def __init__(self,genes,individuals_length,decode,fitness, total_distance):\n",
    "        self.genes= genes\n",
    "        self.individuals_length= individuals_length\n",
    "        self.decode= decode\n",
    "        self.fitness= fitness\n",
    "        self.total_distance= total_distance\n",
    "\n",
    "    def mutation(self, chromosome, prob):\n",
    "            def split_at_values(lst, values):\n",
    "                indices = [i for i, x in enumerate(lst) if x in values]\n",
    "                size = len(lst) \n",
    "                res = [lst[i: j] for i, j in zip([0] + [x+1 for x in indices], indices + ([size] if indices[-1] != size else []))]\n",
    "                return res, indices\n",
    "            \n",
    "            def inversion_mutation(chromosome_aux):\n",
    "                chromosome = chromosome_aux\n",
    "                \n",
    "                index1 = randrange(0,len(chromosome))\n",
    "                index2 = randrange(index1,len(chromosome))\n",
    "                \n",
    "                chromosome_mid = chromosome[index1:index2]\n",
    "                chromosome_mid.reverse()\n",
    "                \n",
    "                chromosome_result = chromosome[0:index1] + chromosome_mid + chromosome[index2:]\n",
    "                \n",
    "                return chromosome_result\n",
    "            print(chromosome)\n",
    "            vehiclesTripList, indices = split_at_values(chromosome, vehicles)\n",
    "            i = 0\n",
    "            for vehicleTrip in vehiclesTripList:\n",
    "                if len(vehicleTrip) >= 2:\n",
    "                    if random.random() < prob :\n",
    "                        aux = inversion_mutation(vehicleTrip)\n",
    "                        chromosome[indices[i]+1:indices[i+1]] = aux\n",
    "                i+=1\n",
    "            print(chromosome)\n",
    "            return chromosome\n",
    "\n",
    "    def crossover(self,parent1, parent2):\n",
    "        def process_vehicle_repeated(copy_child1,copy_child2):\n",
    "            count1=0\n",
    "            for gen1 in copy_child1[:pos]:\n",
    "                if gen1 in vehicles:\n",
    "                    repeat = 0\n",
    "                    repeat = copy_child1.count(gen1)\n",
    "                    if repeat > 1:#If need to fix repeated gen\n",
    "                        count2=0\n",
    "                        for gen2 in parent1[pos:]:#Choose next available gen\n",
    "                            if gen2 not in vehicles:\n",
    "                                child1[count1] = parent1[pos:][count2]\n",
    "                                break\n",
    "                            count2+=1\n",
    "                count1+=1\n",
    "            \n",
    "            count1=0\n",
    "            for gen1 in copy_child2[:pos]:\n",
    "                if gen1 in vehicles:\n",
    "                    repeat = 0\n",
    "                    repeat = copy_child2.count(gen1)\n",
    "                    if repeat > 1:#If need to fix repeated gen\n",
    "                        count2=0\n",
    "                        for gen2 in parent2[pos:]:#Choose next available gen\n",
    "                            if gen2 not in vehicles:\n",
    "                                child2[count1] = parent2[pos:][count2]\n",
    "                                break\n",
    "                            count2+=1\n",
    "                count1+=1\n",
    "            \n",
    "            return [child1,child2]\n",
    "        \n",
    "        \n",
    "        def process_gen_repeated_in_trip(copy_child1,copy_child2):\n",
    "            def split_at_values(lst, values):\n",
    "                indices = [i for i, x in enumerate(lst) if x in values]\n",
    "                size = len(lst) \n",
    "                res = [lst[i: j] for i, j in zip([0] + [x+1 for x in indices], indices + ([size] if indices[-1] != size else []))]\n",
    "                return res, indices\n",
    "            vehiclesTripList, indices = split_at_values(copy_child1, vehicles)\n",
    "            i=0\n",
    "            for vehicleTrip in vehiclesTripList:\n",
    "                if len(vehicleTrip) != 0:\n",
    "                    count1=0\n",
    "                    for gen1 in vehicleTrip:\n",
    "                        repeat = 0\n",
    "                        repeat = vehicleTrip.count(gen1)\n",
    "                        if repeat > 1:#If need to fix repeated gen\n",
    "                            count2=0\n",
    "                            for gen2 in parent1[pos:]:#Choose next available gen\n",
    "                                if gen2 not in vehicleTrip and gen2 not in vehicles:\n",
    "                                    vehicleTrip[count1] = parent1[pos:][count2]\n",
    "                                    break\n",
    "                                count2+=1\n",
    "                        count1+=1\n",
    "                print(indices)\n",
    "                print(child1)\n",
    "                child1[indices[i]+1:indices[i+1]] = vehicleTrip\n",
    "                i += 1\n",
    "                \n",
    "            vehiclesTripList, indices = split_at_values(copy_child2, vehicles)\n",
    "            i = 0\n",
    "            for vehicleTrip in vehiclesTripList:\n",
    "                if len(vehicleTrip) != 0:\n",
    "                    count1=0\n",
    "                    for gen1 in vehicleTrip:\n",
    "                        repeat = 0\n",
    "                        repeat = vehicleTrip.count(gen1)\n",
    "                        if repeat > 1:#If need to fix repeated gen\n",
    "                            count2=0\n",
    "                            for gen2 in parent2[pos:]:#Choose next available gen\n",
    "                                if gen2 not in vehicleTrip and gen2 not in vehicles:\n",
    "                                    vehicleTrip[count1] = parent2[pos:][count2]\n",
    "                                    break\n",
    "                                count2+=1\n",
    "                        count1+=1\n",
    "                \n",
    "                child2[indices[i]+1:indices[i+1]] = vehicleTrip\n",
    "                i += 1\n",
    "                        \n",
    "            return [child1,child2]\n",
    "            \n",
    "        if min(len(parent1), len(parent2)) > 2:\n",
    "            \n",
    "            pos=random.randrange(1, min(len(parent1), len(parent2)))\n",
    "            child1 = parent1[:pos] + parent2[pos:] \n",
    "            child2 = parent2[:pos] + parent1[pos:] \n",
    "            list_childs = process_vehicle_repeated(child1, child2)\n",
    "            list_childs1 = process_gen_repeated_in_trip(list_childs[0], list_childs[1])\n",
    "            #list_childs2 = process_vehicle_absent(list_childs1[0], list_childs1[1])\n",
    "           # print(list_childs2[0])\n",
    "           # print(list_childs2[1])\n",
    "            return list_childs1\n",
    "        else:\n",
    "            return [parent1, parent2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "def decodeVRP(chromosome):    \n",
    "    decoded_list=[]\n",
    "    for k in chromosome:\n",
    "        if k in vehicles:\n",
    "            decoded_list.append(frontier)\n",
    "        else:\n",
    "            decoded_list.append(k)\n",
    "    return decoded_list\n",
    "\n",
    "def fitnessVRP(chromosome):\n",
    "    def split_at_values(lst, values):\n",
    "        indices = [i for i, x in enumerate(lst) if x in values]\n",
    "        size = len(lst) \n",
    "        res = [lst[i: j] for i, j in zip([0] + [x+1 for x in indices], indices + ([size] if indices[-1] != size else []))]\n",
    "        return res\n",
    "\n",
    "    fitness_value = 0\n",
    "    # methode1 : finettes s the sum of distances in a chromosome\n",
    "    #for i in range(0, len(chromosome)-1):\n",
    "    #    if chromosome[i] not in vehicles and chromosome[i+1] not in vehicles:\n",
    "    #        fitness_value += distances[chromosome[i], chromosome[i+1]]\n",
    "    print(chromosome)\n",
    "    \n",
    "    \n",
    "    vehiclesTripList = split_at_values(chromosome, vehicles)\n",
    "    # methode 2 : fitness is max distance of trips\n",
    "    tripDistances = []\n",
    "    for vehicleTrip in vehiclesTripList:\n",
    "        \n",
    "        distance = 0\n",
    "        nbStops = len(vehicleTrip)\n",
    "        if nbStops != 0:\n",
    "            for i in range(0, len(vehicleTrip)-1): \n",
    "                distance += distances[vehicleTrip[i], vehicleTrip[i+1]]\n",
    "            tripDistances.append(distance)\n",
    "        else:\n",
    "            tripDistances.append(distance)\n",
    "    fitness_value = max(tripDistances)\n",
    "\n",
    "    for vehicleTrip in vehiclesTripList:\n",
    "        nbStops = len(vehicleTrip)\n",
    "        if nbStops != 0:\n",
    "            \n",
    "            # trip must start by Home:\n",
    "            if 'W' in vehicleTrip[0]:\n",
    "                fitness_value += 100000\n",
    "            # each trip must contain Hs and Ws\n",
    "            H_exist = False\n",
    "            W_exist = False\n",
    "            for stop in vehicleTrip:\n",
    "                if \"W\" in stop:\n",
    "                    W_exist = True\n",
    "                if \"H\" in stop:\n",
    "                    H_exist = True\n",
    "            if H_exist == False or W_exist == False:\n",
    "                fitness_value += 100000\n",
    "                    \n",
    "            # if W is before H this is not a valid chromosome, should have a very big distance (bad fitness)\n",
    "            for i in range(0, nbStops-1):\n",
    "                if \"W\" in vehicleTrip[i] and \"H\" in vehicleTrip[i+1]:\n",
    "                    fitness_value += 100000\n",
    "                    break\n",
    "        \n",
    "    return fitness_value\n",
    "\n",
    "def totalDistanceVRP(chromosome):\n",
    "    def split_at_values(lst, values):\n",
    "        indices = [i for i, x in enumerate(lst) if x in values]\n",
    "        size = len(lst) \n",
    "        res = [lst[i: j] for i, j in zip([0] + [x+1 for x in indices], indices + ([size] if indices[-1] != size else []))]\n",
    "        return res\n",
    "\n",
    "    total_distance = 0\n",
    "    \n",
    "    for i in range(0, len(chromosome)-1):\n",
    "        if chromosome[i] not in vehicles and chromosome[i+1] not in vehicles:\n",
    "            total_distance += distances[chromosome[i], chromosome[i+1]]\n",
    "    \n",
    "    return total_distance\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#========================================================== FIRST PART: GENETIC OPERATORS============================================\n",
    "# Here We defined the requierements functions that the GA needs to work \n",
    "# The function receives as input:\n",
    "# * problem_genetic: an instance of the class Problem_Genetic, with\n",
    "#     the optimization problem that we want to solve.\n",
    "# * k: number of participants on the selection tournaments.\n",
    "# * opt: max or min, indicating if it is a maximization or a\n",
    "#     minimization problem.\n",
    "# * ngen: number of generations (halting condition)\n",
    "# * size: number of individuals for each generation\n",
    "# * ratio_cross: portion of the population which will be obtained by\n",
    "#     means of crossovers. \n",
    "# * prob_mutate: probability that a gene mutation will take place.\n",
    "#=====================================================================================================================================\n",
    "\n",
    "\n",
    "def genetic_algorithm_t(Problem_Genetic,k,opt,ngen,populationSize,ratio_cross,prob_mutate):\n",
    "    \n",
    "    def initial_population(Problem_Genetic,populationSize):   \n",
    "        def generate_chromosome():\n",
    "            chromosome=[]\n",
    "            for vehicle in vehicles:\n",
    "                \n",
    "                vehicle_stops_nb = random.randrange(0,len(all_stations))\n",
    "                \n",
    "                trip = random.sample(list(all_stations), vehicle_stops_nb)\n",
    "                chromosome += [vehicle]\n",
    "                chromosome += trip\n",
    "            print(chromosome)\n",
    "            return chromosome\n",
    "        return [generate_chromosome() for _ in range(populationSize)]\n",
    "            \n",
    "    def new_generation_t(Problem_Genetic,k,opt,population,n_parents,n_directs,prob_mutate):\n",
    "        \n",
    "        def tournament_selection(Problem_Genetic,population,n,k,opt):\n",
    "            winners=[]\n",
    "            for _ in range(n):\n",
    "                \n",
    "                elements = random.sample(population,k)\n",
    "                winners.append(opt(elements,key=Problem_Genetic.fitness))\n",
    "            return winners\n",
    "        \n",
    "        def cross_parents(Problem_Genetic,parents):\n",
    "            childs=[]\n",
    "            for i in range(0,len(parents),2):\n",
    "                childs.extend(Problem_Genetic.crossover(parents[i],parents[i+1]))\n",
    "            return childs\n",
    "    \n",
    "        def mutate(Problem_Genetic,population,prob):\n",
    "            for i in population:\n",
    "                Problem_Genetic.mutation(i,prob)\n",
    "            return population\n",
    "        \n",
    "        print(\"##############################directs\")\n",
    "        directs = tournament_selection(Problem_Genetic, population, n_directs, k, opt)\n",
    "        print(\"##############################Crosses\")\n",
    "        crosses = cross_parents(Problem_Genetic,\n",
    "                                tournament_selection(Problem_Genetic, population, n_parents, k, opt))\n",
    "        #print(\"##############################mutattion\")\n",
    "        #mutations = mutate(Problem_Genetic, crosses, prob_mutate)\n",
    "        new_generation = directs + crosses\n",
    "        \n",
    "        return new_generation\n",
    "    \n",
    "    population = initial_population(Problem_Genetic, populationSize)\n",
    "    n_parents = round(populationSize*ratio_cross)\n",
    "    n_parents = (n_parents if n_parents%2==0 else n_parents-1)\n",
    "    n_directs = populationSize - n_parents\n",
    "    print(\"n_of parent to cross \", n_parents)\n",
    "    for _ in range(ngen):\n",
    "        population = new_generation_t(Problem_Genetic, k, opt, population, n_parents, n_directs, prob_mutate)\n",
    "    \n",
    "    bestChromosome = opt(population, key = Problem_Genetic.fitness)\n",
    "    #print(\"Chromosome: \", bestChromosome)\n",
    "    genotype = Problem_Genetic.decode(bestChromosome)\n",
    "    print (\"Solution: {0}, Fitness: {1}, TotalDistance: {2}\".format(genotype,Problem_Genetic.fitness(bestChromosome), Problem_Genetic.total_distance(bestChromosome)))\n",
    "    return (genotype,Problem_Genetic.fitness(bestChromosome))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 2, 3]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "poop = []\n",
    "poop += [1,2,3]\n",
    "poop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "inputFileName = \"data/data_transportationPb_mini.xlsx\"\n",
    "param_data = pd.read_excel(inputFileName, \"Param\", skiprows=0)\n",
    "trips_df = pd.read_excel(inputFileName, \"Trips\", skiprows=0)\n",
    "distances_df = pd.read_excel(inputFileName, \"Distance\", skiprows=0)\n",
    "\n",
    "nbVehicles = param_data.loc[param_data[\"Param\"] == \"NbVehicles\"][[\"Value1\",\"Value2\",\"Value3\"]].sum(axis=1).values[0]\n",
    "nbHome = param_data.loc[param_data[\"Param\"] == \"NbHome\"][\"Value1\"].values[0]\n",
    "nbWork = param_data.loc[param_data[\"Param\"] == \"NbWork\"][\"Value1\"].values[0]\n",
    "timePerStation = param_data.loc[param_data[\"Param\"] == \"TimePerStation (min)\"][\"Value1\"].values[0]\n",
    "speed = param_data.loc[param_data[\"Param\"] == \"Speed (min/km)\"][\"Value1\"].values[0]\n",
    "trips = [tuple(x) for x in trips_df[[\"Departure\", \"Arrival\"]].dropna().to_numpy()]\n",
    "all_stations = np.concatenate((trips_df[\"Departure\"].dropna().unique(), trips_df[\"Arrival\"].dropna().unique()))\n",
    "distances = dict()\n",
    "for station_i in all_stations:\n",
    "    for station_j in all_stations:\n",
    "        if station_i != station_j:\n",
    "            row = distances_df[(distances_df[\"Departure\"] == station_i) & (distances_df[\"Arrival\"] == station_j)]\n",
    "            if len(row) != 0:\n",
    "                distances[station_i, station_j] = row[\"Distance\"].values[0]\n",
    "                distances[station_j, station_i] = row[\"Distance\"].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['H1', 'H2', 'H3', 'H4', 'W1', 'W2', 'W3'], dtype=object)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_stations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EXECUTING  1  INSTANCES \n",
      "---------------------------------------------------------Executing FIRST PART: VRP --------------------------------------------------------- \n",
      "\n",
      "Frontier =  ######\n",
      "\n",
      "['vehicle0', 'H3', 'W3', 'W2', 'H4', 'H2', 'vehicle1', 'H3', 'H2', 'W1']\n",
      "['vehicle0', 'H3', 'W3', 'H1', 'vehicle1', 'H1', 'H4', 'H2', 'H3', 'W2', 'W3']\n",
      "['vehicle0', 'H3', 'vehicle1', 'W1', 'W3', 'H1']\n",
      "['vehicle0', 'H4', 'H3', 'W1', 'vehicle1', 'H2']\n",
      "['vehicle0', 'H1', 'H2', 'H4', 'W1', 'W2', 'W3', 'vehicle1', 'H4', 'H1', 'H3', 'W2', 'H2']\n",
      "['vehicle0', 'W3', 'H1', 'H3', 'H4', 'vehicle1', 'H1', 'H3']\n",
      "['vehicle0', 'W3', 'vehicle1', 'H1', 'W2', 'W1', 'W3', 'H2', 'H3']\n",
      "['vehicle0', 'W1', 'vehicle1', 'W3', 'W1']\n",
      "['vehicle0', 'W2', 'W1', 'H2', 'vehicle1', 'H4', 'H3', 'W2']\n",
      "['vehicle0', 'H2', 'H1', 'H3', 'W1', 'W3', 'H4', 'vehicle1', 'H4']\n",
      "n_of parent to cross  8\n",
      "##############################directs\n",
      "['vehicle0', 'H3', 'vehicle1', 'W1', 'W3', 'H1']\n",
      "['vehicle0', 'W3', 'H1', 'H3', 'H4', 'vehicle1', 'H1', 'H3']\n",
      "['vehicle0', 'H3', 'vehicle1', 'W1', 'W3', 'H1']\n",
      "['vehicle0', 'W3', 'vehicle1', 'H1', 'W2', 'W1', 'W3', 'H2', 'H3']\n",
      "##############################Crosses\n",
      "['vehicle0', 'W2', 'W1', 'H2', 'vehicle1', 'H4', 'H3', 'W2']\n",
      "['vehicle0', 'H3', 'vehicle1', 'W1', 'W3', 'H1']\n",
      "['vehicle0', 'H4', 'H3', 'W1', 'vehicle1', 'H2']\n",
      "['vehicle0', 'W2', 'W1', 'H2', 'vehicle1', 'H4', 'H3', 'W2']\n",
      "['vehicle0', 'W3', 'vehicle1', 'H1', 'W2', 'W1', 'W3', 'H2', 'H3']\n",
      "['vehicle0', 'H3', 'W3', 'H1', 'vehicle1', 'H1', 'H4', 'H2', 'H3', 'W2', 'W3']\n",
      "['vehicle0', 'H4', 'H3', 'W1', 'vehicle1', 'H2']\n",
      "['vehicle0', 'W3', 'H1', 'H3', 'H4', 'vehicle1', 'H1', 'H3']\n",
      "['vehicle0', 'H3', 'W3', 'H1', 'vehicle1', 'H1', 'H4', 'H2', 'H3', 'W2', 'W3']\n",
      "['vehicle0', 'W1', 'vehicle1', 'W3', 'W1']\n",
      "['vehicle0', 'H3', 'W3', 'W2', 'H4', 'H2', 'vehicle1', 'H3', 'H2', 'W1']\n",
      "['vehicle0', 'H2', 'H1', 'H3', 'W1', 'W3', 'H4', 'vehicle1', 'H4']\n",
      "['vehicle0', 'H4', 'H3', 'W1', 'vehicle1', 'H2']\n",
      "['vehicle0', 'W1', 'vehicle1', 'W3', 'W1']\n",
      "['vehicle0', 'W3', 'H1', 'H3', 'H4', 'vehicle1', 'H1', 'H3']\n",
      "['vehicle0', 'H1', 'H2', 'H4', 'W1', 'W2', 'W3', 'vehicle1', 'H4', 'H1', 'H3', 'W2', 'H2']\n",
      "[0, 4]\n",
      "['vehicle0', 'W2', 'W1', 'H2', 'vehicle1', 'H2']\n",
      "[0, 4]\n",
      "['vehicle0', 'vehicle1', 'H2']\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-55-df789c73c2c0>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     28\u001b[0m     \u001b[0mgenetic_problem_instances\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     29\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"EXECUTING \"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mgenetic_problem_instances\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\" INSTANCES \"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 30\u001b[1;33m     \u001b[0mVRP\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mgenetic_problem_instances\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-55-df789c73c2c0>\u001b[0m in \u001b[0;36mVRP\u001b[1;34m(k)\u001b[0m\n\u001b[0;32m     16\u001b[0m         \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Total time: \"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mt1\u001b[0m \u001b[1;33m-\u001b[0m \u001b[0mt0\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m\" secs.\\n\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     17\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 18\u001b[1;33m     \u001b[0mfirst_part_GA\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     19\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     20\u001b[0m \u001b[1;31m#---------------------------------------- AUXILIARY DATA FOR TESTING --------------------------------\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-55-df789c73c2c0>\u001b[0m in \u001b[0;36mfirst_part_GA\u001b[1;34m(k)\u001b[0m\n\u001b[0;32m     10\u001b[0m         \u001b[0mt0\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     11\u001b[0m         \u001b[1;32mwhile\u001b[0m \u001b[0mcont\u001b[0m \u001b[1;33m<=\u001b[0m \u001b[0mk\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 12\u001b[1;33m             \u001b[0mgenetic_algorithm_t\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mVRP_PROBLEM\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m2\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmin\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m2\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m10\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m0.8\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m0.05\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     13\u001b[0m             \u001b[0mcont\u001b[0m\u001b[1;33m+=\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     14\u001b[0m         \u001b[0mt1\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtime\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-51-4fb1523cdc1e>\u001b[0m in \u001b[0;36mgenetic_algorithm_t\u001b[1;34m(Problem_Genetic, k, opt, ngen, populationSize, ratio_cross, prob_mutate)\u001b[0m\n\u001b[0;32m     69\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"n_of parent to cross \"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mn_parents\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     70\u001b[0m     \u001b[1;32mfor\u001b[0m \u001b[0m_\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mngen\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 71\u001b[1;33m         \u001b[0mpopulation\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnew_generation_t\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mProblem_Genetic\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mk\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mopt\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mpopulation\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mn_parents\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mn_directs\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mprob_mutate\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     72\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     73\u001b[0m     \u001b[0mbestChromosome\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mopt\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpopulation\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkey\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mProblem_Genetic\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfitness\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-51-4fb1523cdc1e>\u001b[0m in \u001b[0;36mnew_generation_t\u001b[1;34m(Problem_Genetic, k, opt, population, n_parents, n_directs, prob_mutate)\u001b[0m\n\u001b[0;32m     55\u001b[0m         \u001b[0mdirects\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtournament_selection\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mProblem_Genetic\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mpopulation\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mn_directs\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mk\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mopt\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     56\u001b[0m         \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"##############################Crosses\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 57\u001b[1;33m         crosses = cross_parents(Problem_Genetic,\n\u001b[0m\u001b[0;32m     58\u001b[0m                                 tournament_selection(Problem_Genetic, population, n_parents, k, opt))\n\u001b[0;32m     59\u001b[0m         \u001b[1;31m#print(\"##############################mutattion\")\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-51-4fb1523cdc1e>\u001b[0m in \u001b[0;36mcross_parents\u001b[1;34m(Problem_Genetic, parents)\u001b[0m\n\u001b[0;32m     44\u001b[0m             \u001b[0mchilds\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     45\u001b[0m             \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mparents\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 46\u001b[1;33m                 \u001b[0mchilds\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mextend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mProblem_Genetic\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcrossover\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mparents\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mparents\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     47\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mchilds\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     48\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-49-9be1cf033835>\u001b[0m in \u001b[0;36mcrossover\u001b[1;34m(self, parent1, parent2)\u001b[0m\n\u001b[0;32m    140\u001b[0m             \u001b[0mchild2\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mparent2\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m:\u001b[0m\u001b[0mpos\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mparent1\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mpos\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    141\u001b[0m             \u001b[0mlist_childs\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mprocess_vehicle_repeated\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mchild1\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mchild2\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 142\u001b[1;33m             \u001b[0mlist_childs1\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mprocess_gen_repeated_in_trip\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlist_childs\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlist_childs\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    143\u001b[0m             \u001b[1;31m#list_childs2 = process_vehicle_absent(list_childs1[0], list_childs1[1])\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    144\u001b[0m            \u001b[1;31m# print(list_childs2[0])\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-49-9be1cf033835>\u001b[0m in \u001b[0;36mprocess_gen_repeated_in_trip\u001b[1;34m(copy_child1, copy_child2)\u001b[0m\n\u001b[0;32m    109\u001b[0m                 \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mindices\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    110\u001b[0m                 \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mchild1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 111\u001b[1;33m                 \u001b[0mchild1\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mindices\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[0mindices\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m+\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mvehicleTrip\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    112\u001b[0m                 \u001b[0mi\u001b[0m \u001b[1;33m+=\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    113\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "def VRP(k):\n",
    "    VRP_PROBLEM = Problem_Genetic(list(all_stations)+ vehicles[:-1],\n",
    "                                  len(list(all_stations)), lambda x : decodeVRP(x), lambda y: fitnessVRP(y),lambda y: totalDistanceVRP(y))\n",
    "    \n",
    "    def first_part_GA(k):\n",
    "        cont  = 0\n",
    "        print (\"---------------------------------------------------------Executing FIRST PART: VRP --------------------------------------------------------- \\n\")\n",
    "        print(\"Frontier = \",frontier)\n",
    "        print(\"\")\n",
    "        t0 = time()\n",
    "        while cont <= k: \n",
    "            genetic_algorithm_t(VRP_PROBLEM, 2, min, 2, 10, 0.8, 0.05)\n",
    "            cont+=1\n",
    "        t1 = time()\n",
    "        print(\"\\n\") \n",
    "        print(\"Total time: \",(t1 - t0),\" secs.\\n\")\n",
    "    \n",
    "    first_part_GA(k)\n",
    "\n",
    "#---------------------------------------- AUXILIARY DATA FOR TESTING --------------------------------\n",
    "\n",
    "vehicles = ['vehicle'+str(i) for i in range(0, int(nbVehicles))]\n",
    "frontier = \"######\"\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    # Constant that is an instance object \n",
    "    genetic_problem_instances = 1\n",
    "    print(\"EXECUTING \", genetic_problem_instances, \" INSTANCES \")\n",
    "    VRP(genetic_problem_instances)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['H1', 'H2', 'H3', 'H4', 'W1', 'W2', 'W3'], dtype=object)"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_stations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "lst = ['vehicle0', 'H2', 'vehicle1', 'H2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
